src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp

```diff
@@ -179,57 +70,68 @@ APPLICATION_INFO::EnsureApplicationCreated(
     STRU                struHostFxrDllLocation;
     STACK_STRU(struFileName, 300);  // >MAX_PATH
 
-    if (m_pApplication != NULL)
+    if (m_pApplication != nullptr && m_pApplication->QueryStatus() != OFFLINE)
     {
         return S_OK;
     }
 
-    // one optimization for failure scenario is to reduce the lock scope
     SRWExclusiveLock lock(m_srwLock);
 
+    if (m_pApplication != nullptr)
     {
-        if (m_pApplication != NULL)
+        if (m_pApplication->QueryStatus() == OFFLINE)
         {
-            // another thread created the applicaiton
-            FINISHED(S_OK);
+            LOG_INFO("Application went offline");
+            // Application that went offline
+            // are supposed to recycle themselves
+            m_pApplication->DereferenceApplication();
```

Comment:
There is a race condition here with the app being deleted while there is an incoming request. After calling ensure application created, we call ReferenceApplication. However, between that time, app offline could be dropped and we dereference the application, causing the request which just released this lock to AV. I would call reference application inside of EnsureApplicationCreated to avoid this issue.

Reply:
ExtractApplication has a lock too and calls ReferenceApplication under it:

https://github.com/aspnet/IISIntegration/blob/690a95dc09ac0a081347cde9effed36ce47495ca/src/AspNetCoreModuleV2/AspNetCore/Inc/applicationinfo.h#L125

Reply:
but between the time it unlocks and locks, state can change. I think we just need to call extract application inside of GetORCreateApplicationInfo.